# Blatt 6
---

In [ ]:
buch = "a"
for i in range(26):
    print(f"{i} : {buch}")
    buch = chr(ord(buch) + 1)

## Aufgabe 21)

### a)

(N, e) = (273373, 65537)

Verschlüsseln von "bielefeld"

Unterteilung in 3 messages: "bie" "lef" "eld"

$m_1$ = 10804

$m_2$ = 110405

$m_3$ = 41103

Verschlüsselung mit: $m^e$ mod $N$

In [20]:
import numpy as np

m1 = 10804
m2 = 110405
m3 = 41103
N = 273373
e = 65537

print(np.mod(m1**e, N))
print(np.mod(m2**e, N))
print(np.mod(m3**e, N))

246231
17787
73173


246231 = ykf
17787 = bzj
73173 = hfv

Die verschlüsselte Botschaft lautet: __ykfbzjhfv__

### b)

Entschlüsseln Sie die Botschaft (0, 149578, 35224)

Ansatz: Mithilfe von Primfaktorzerlegung p und q bestimmen um dann den geheimen Schlüssel von Alice zu berechnen:

In [17]:
def primfaktorzerlegung(zahl):
    faktoren = []
    i = 2
    while zahl > 1:
        if zahl % i == 0:
            faktoren.append(i)
            zahl = zahl / i
        else:
            i += 1
    return faktoren

print(primfaktorzerlegung(273373))

[461, 593]


p = 461, q=593, $\phi(N)$ = 460*592 = 272320

Da wir nun $\phi(N)$ haben können wir Mithilfe des erweiterten euklidischen Algorithmus d bestimmen damit wir mit den geheimen Schlüssel die Nachricht entschlüsseln können. 

a         b         q         r         x          y
272320
...


d = 15233

Nun können wir mithilfe der Entschlüsselungsformel die Nachricht entschlüsseln

$m = c^d mod N$

$0^{15233}$ = $0$ mod $273373$ = $0$

$149578^{15233}$ mod $273373$ = $12481$

$35224^{15233}$ mod $273373$ = $4758$

Die Nachricht:

__abydvg__

In [32]:
np.mod(58, 26)

6

## Aufgabe 22)

### a)

(N, e) = (1007, 31)

$\phi(N)$ = 936

0 = $−q^2 + (N − ϕ(N) + 1)q − N$

0 = $-q^2$ + (1007 - 936 + 1)q - 1007

0 = $-q^2$ + 72q - 1007

$q^2$ - 72q + 1007 = 0

$p-q$ Formel:

$\frac{72}{2} +- \sqrt{\frac{-72}{2}^2 - 1007}$ = $ 36 +- \sqrt{36^2 - 1007}$

In [15]:
print(36 + np.sqrt(36**2 - 1007))
print(36 - np.sqrt(36**2 - 1007))

53.0
19.0


Dabei ist:

$q_1$ = 53

$q_2$ = 19

Nun können wir $p_1$ und $p_2$ ausrechnen.

$p_1$ = 1007 / 53 = 19

$p_2$ = 1007 / 19 = 53

Dadurch wissen wir nun das p und q die Zahlen 19 und 53 sind.

### b)

c = 358

(N, e) = (1007, 31)

p = 19, q = 53

$\phi(N)$ = 18*52 = 936

Ansatz: Geheimen Schlüssel d mithilfe des erweiterten Euklidischen Algorithmus berechnen

31 = 0 * 936 + 31  

936 = 30*31 + 6     

31 = 5*6 + 1         

6 = 6*1               

0   1

1   0

-30 1

151 -5

d = $151$

$m = c^d$ mod $N$

$m = 358^{151}$ mod $1007$

$m = 17$

$m = r$

In [34]:
np.mod(358**151, 1007)

17

## Aufgabe 23

### a)

$f(e, m_{\scriptstyle{1}}*m_{\scriptstyle{2}}) = f(e, m_{\scriptstyle{1}}) * f(e, m_{\scriptstyle{2}})$

$f(e, m_{\scriptstyle{1}}) * f(e, m_{\scriptstyle{2}}) = m_{\scriptstyle{1}}^e mod N * m_{\scriptstyle{2}}^e mod N$

Modulare Gleichheit der Multiplikation:

$m_{\scriptstyle{1}}^e mod N * m_{\scriptstyle{2}}^e mod N = (m_{\scriptstyle{1}}^e * m_{\scriptstyle{2}}^e) mod N$

$(m_{\scriptstyle{1}}^e * m_{\scriptstyle{2}}^e) mod N = (m_{\scriptstyle{1}} * m_{\scriptstyle{2}})^e mod N$

$(m_{\scriptstyle{1}} * m_{\scriptstyle{2}})^e mod N = \underline{\underline{f(e, m_{\scriptstyle{1}} * m_{\scriptstyle{2}})}}$

### b)

aus a) $f(e, m_{\scriptstyle{1}}*m_{\scriptstyle{2}}) = f(e, m_{\scriptstyle{1}}) * f(e, m_{\scriptstyle{2}})$

Und es gilt:
$c_{\scriptstyle{1}} = f(e, m_{\scriptstyle{1}})$

Wenn Eve einen beliebigen Text $m_{\scriptstyle{2}}$ von Alice entschlüsseln lässt und diesen erhält, weiß Eve: 

$c_{\scriptstyle{2}} = f(e, m_{\scriptstyle{2}})$ und $m_{\scriptstyle{2}}$ ist der entschlüsselte Text

Somit sind $c_{\scriptstyle{1}}$ und $c_{\scriptstyle{2}}$ und $m_{\scriptstyle{2}}$ bekannt.

Es gilt:

$f(e, m_{\scriptstyle{1}}) * f(e, m_{\scriptstyle{2}}) = c_{\scriptstyle{1}} * c_{\scriptstyle{2}}$

$f(e, m_{\scriptstyle{1}}*m_{\scriptstyle{2}}) = c_{\scriptstyle{1}} * c_{\scriptstyle{2}} = (f(e, m_{\scriptstyle{1}})) * (f(e, m_{\scriptstyle{2}}))$

$f(e, m_{\scriptstyle{1}}*m_{\scriptstyle{2}}) = \frac{c_{\scriptstyle{1}} * c_{\scriptstyle{2}}}{f(e, m_{\scriptstyle{2}})}$

Dadurch, dass die multiplikative Gleichheit bei der modulo Operation gilt, kann man durch $m_{\scriptstyle{2}}$ teilen und man erhält folgendes:

$m_{\scriptstyle{1}}^e mod N * m_{\scriptstyle{2}}^e mod N = \frac{c_{\scriptstyle{1}}* c_{\scriptstyle{2}}}{m_{\scriptstyle{2}}^e mod N}$

Nebenrechnung:
$\frac{m_{\scriptstyle{1}}^e mod N * m_{\scriptstyle{2}}^e mod N}{m_{\scriptstyle{2}}^e mod N} = m_{\scriptstyle{1}}^e$

$\underline{\underline{
m_{\scriptstyle{1}} = \frac{c_{\scriptstyle{1}} * c_{\scriptstyle{2}}}{f(e, m_{\scriptstyle{2}})*m_{\scriptstyle{2}}}
}}$




### c)

Übergabe der Schlüssel bereits verschlüsselt.

## Aufgabe 24)

N = pq mit p, q Primzahlen

### a)

$\frac{p+q}{2} \in \mathbb{N}$ für p > q > 2

Da p und q Primzahlen und größer 2 sind gilt, dass p und q ungerade. Das Produkt zweier ungerader Zahlen ist eine gerade Zahl, welche durch die Beschränkung p, q > 2 auch > 0 ist. Da eine gerade, positive Zahl, geteilt durch eine positive gerade Zahl auch eine positive gerade Zahl ist, gilt:

$\frac{p+q}{2} \in \mathbb{N}$

---

N = $(\frac{p+q}{2})^2 - (\frac{p-q}{2})^2$

N = $(\frac{p+q}{2})*(\frac{p+q}{2}) - (\frac{p-q}{2})*(\frac{p-q}{2})$

N = $(\frac{p^2 + 2pq + q^2}{4}) - (\frac{p^2 - 2pq + q^2}{4})$

N = $\frac{(p^2 + 2pq + q^2) - (p^2 - 2pq + q^2)}{4}$

N = $\frac{4pq}{4}$

$\underline{\underline{N = pq}}$

### b)

N = pq.

Zuerst wird eine natürliche Zahl a gesucht, sodass $\lceil \sqrt{N} \rceil = a^2-N$ eine Quadratzahl ist, das bedeutet, dass dieses a eine ganz Zahl ist und sehr nah an N/2 liegt. Dies liegt daran, dass a somit zwischen p und q liegt und nicht teilerfremd von N sind.

Wenn nun $b = \sqrt{a^2-N}$ gilt, wissen wir, dass b eine kleine Zahl ist.

Wenn a-b und a+b Teiler von N sind, kann davon ausgegangen werden, dass a etwa in der Mitte zwischen p und q liegt. Mit b wurde eine "Schranke" berechnet, die auf beide Seiten von a angewendet wird, um die beiden Zahlen p und q zu berechnen.

Dies ist Sinnvoll, da b klein ist und p und q sehr nah beieinander liegen und $\sqrt{a^2-N}$ einen sehr kleinen Teiler von N darstellt.

p und q werden somit von a+b und a-b repräsentiert.

### c)

N = pq = 294867870917365576583920008771513215081

a = 17171717180217172000

In [ ]:
import math

def getPandQ(a, N):
    print(f"a = {a}")

    if not checkQuadratic(a, N):
        print(f"{a} ist nicht quadratisch")
        a = a+1
        print("---------------")
        getPandQ(a, N)

    b = math.sqrt((a**2)-N)
    print(f"b = {b}")

    p = a-b
    q = a+b

    if (N % p == 0) and (N % q == 0):
        print(f"p = {p}")
        print(f"q = {q}")
        return p, q
    
    print("p oder q kein Teiler von N")
    a = a+1
    print("---------------")
    getPandQ(a, N)

    

def checkQuadratic(a, N):
    test = (a**2)-N

    return(math.isqrt(test)**2 == test)



tempN = 294867870917365576583920008771513215081
tempA = math.isqrt(tempN) + 1
getPandQ(tempA, tempN)


Python Kernel wird abgebrochen da zu viele Berechnungen durchgeführt werden müssen. 